### PR TITLE
feature: impl graphql relay resolver for user and group

### PR DESCRIPTION
**Pull request by fregataa**
Implement async compatible graphql relay Node and Connection types.
Implement user/group relay Node and Connection.

# Why re-invent GraphQL relay?
Current Graphene(v 3.3.0) has a relay implementation but it does not support asynchronous resolver functions. Additionally, Graphene Connection resolver is implemented to only accept one complete array(`Iterable` value) without considering pagination, which is a huge performance issue.
Here, the newly "invented" Connection took pagination into account. But, pagination order of this Connection can only be either forward or backward in a single query, meaning that queries such as `first` and `last` together or `after` and `before` together are limited.

# Query all groups
```gql
# query all groups
query GroupsQuery {
  group_nodes {
    edges {
      node {
        id
        name
      }
    }
  }
}
```

# Query group with relay id
```gql
query GroupsQuery($id: String!) {
    group_node(id: $id) {
        id
        name
        user_nodes(order: "created_at", offset: 2, first: 3) {
            edges {
                node {
                    id
                    username
                    created_at
                }
            }
            count
            pageInfo {
                hasNextPage
                hasPreviousPage
            }
        }
    }
}

# variables
# {
#    "id": "R3JvdXBOb2RlOjJkZTJiOTY5LTFkMDQtNDhhNi1hZjE2LTBiYzhhZGIzYzgzMQ=="
# }
```

# Query node with node's global id
```gql
{
    node(id: "R3JvdXBOb2RlOjJkZTJiOTY5LTFkMDQtNDhhNi1hZjE2LTBiYzhhZGIzYzgzMQ==") {
        ... on GroupNode {
            name
            description
        }
    }
}

```

**Checklist:** (if applicable)

- [x] Milestone metadata specifying the target backport version
- [ ] Documentation
- [ ] API server-client counterparts (e.g., manager API -> client SDK)
- [ ] Test case(s) to demonstrate the difference of before/after

### CODE REVIEW

**Comment by yomybaby** [comment]
It works properly.
```graphql
  group_nodes(order:"-name") {
    edges {
      node {
        id
        name
        user_nodes(first:1) {
          edges {
            node {
              username
            }
          }
          count
        }
      }
      cursor
    }
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
  }
  ```

**Review by yomybaby — APPROVED** [review]
👍

**Review by achimnol — CHANGES_REQUESTED** [review]
Just a minor refactoring to make it more readable:
`SQLInfoForGQLConn` → `GraphQLConnectionSQLInfo`

File: src/ai/backend/manager/models/base.py [hunk 1]
```diff
@@ -1098,3 +1106,224 @@ class InferenceSessionErrorInfo(graphene.ObjectType):
...
+
+def _build_sql_stmt_from_connection_arg(
+    info: graphene.ResolveInfo,
+    orm_class,
+    id_column: sa.Column,
+    filter_expr: str | None = None,
+    order_expr: str | None = None,
+    *,
+    connection_arg: ConnectionArgs,
+) -> tuple[sa.sql.Select, list[WhereClauseType]]:
```
Comment (rapsealk) [inline src/ai/backend/manager/models/base.py]:
```suggestion
def _build_sql_stmt_from_connection_args(
    info: graphene.ResolveInfo,
    orm_class,
    id_column: sa.Column,
    filter_expr: str | None = None,
    order_expr: str | None = None,
    *,
    connection_args: ConnectionArgs,
) -> tuple[sa.sql.Select, list[WhereClauseType]]:
```

File: src/ai/backend/manager/models/base.py [hunk 2]
```diff
@@ -1098,3 +1106,224 @@ class InferenceSessionErrorInfo(graphene.ObjectType):
...
+    stmt = sa.select(orm_class)
+    conditions: list[WhereClauseType] = []
+
+    cursor_id, pagination_order, requested_page_size = connection_arg
```
Comment (rapsealk) [inline src/ai/backend/manager/models/base.py]:
```suggestion
    cursor_id, pagination_order, requested_page_size = connection_args
```

File: src/ai/backend/manager/models/base.py [hunk 3]
```diff
@@ -1098,3 +1106,224 @@ class InferenceSessionErrorInfo(graphene.ObjectType):
...
+            order_expr,
+            connection_arg=connection_arg,
+        )
+        return GraphQLConnectionSQLInfo(
+            stmt,
+            conditions,
+            connection_arg.cursor,
+            connection_arg.pagination_order,
+            connection_arg.requested_page_size,
+        )
```
Comment (rapsealk) [inline src/ai/backend/manager/models/base.py]:
```suggestion
        connection_args = validate_connection_args(
            after=after, first=first, before=before, last=last
        )
        stmt, conditions = _build_sql_stmt_from_connection_arg(
            info,
            orm_class,
            id_column,
            filter_expr,
            order_expr,
            connection_args=connection_args,
        )
        return GraphQLConnectionSQLInfo(
            stmt,
            conditions,
            connection_args.cursor,
            connection_args.pagination_order,
            connection_args.requested_page_size,
        )
```

File: src/ai/backend/manager/models/gql.py [hunk 4]
```diff
@@ -292,6 +303,11 @@ class Queries(graphene.ObjectType):
         is_active=graphene.Boolean(),
     )
 
+    group_node = graphene.Field(
+        GroupNode, id=graphene.String(required=True), description="Added in 24.03.0."
+    )
```
Comment (rapsealk) [inline src/ai/backend/manager/models/gql.py:308]: If `AsyncNode` is to handle Relay Global ID, how about setting `id` field to `AsyncNode` itself?

**Review by rapsealk — COMMENTED** [review]
Minor fixes and suggestions. Thanks for the job!

File: src/ai/backend/manager/models/gql.py (hunk 4 above)
Comment (fregataa) [reply src/ai/backend/manager/models/gql.py:308]: I have tried it, but it needs so much work. so I deprioritized this task

Comment (rapsealk) [reply src/ai/backend/manager/models/gql.py:308]: Let's try this in next PR!

**Review by rapsealk — APPROVED** [review]
LGTM!
I'm wondering if there can be a better name than `OrderingItem` to emphasize the action "ordering".

**Pull request merged**